Leanplum-SDK/Classes/LPRequestStorage.m

```diff
@@ -34,54 +34,115 @@ - (id)init
+        NSMutableArray *requests = [NSMutableArray array];
+        NSInteger count = [_defaults integerForKey:LEANPLUM_DEFAULTS_COUNT_KEY];
+        if (count) {
+            for (NSInteger i = 0; i < count; i++) {
+                NSString *itemKey = [LPRequestStorage itemKeyForIndex:i];
+                NSDictionary *requestArgs = [_defaults dictionaryForKey:itemKey];
+                if (requestArgs) {
+                    [requests addObject:requestArgs];
                 }
+                [_defaults removeObjectForKey:itemKey];
             }
-            NSData *requestData = [NSData dataWithContentsOfFile:self.documentsFilePath];
-            if (requestData) {
-                error = nil;
-                NSMutableArray *savedRequests = [NSPropertyListSerialization
-                                                 propertyListWithData:requestData
-                                                 options:NSPropertyListMutableContainers
-                                                 format:NULL error:&error];
-                if (savedRequests) {
-                    requests = savedRequests;
-                }
+            [_defaults removeObjectForKey:LEANPLUM_DEFAULTS_COUNT_KEY];
+            [_defaults synchronize];
+            if (![[NSFileManager defaultManager] fileExistsAtPath:self.documentsFilePath]) {
+                [self saveRequests:requests];
             }
         }
     }
+}
 
-    LPLog(LPDebug, @"Loaded %lu requests", (unsigned long) requests.count);
+/**
+ * Save requests to documents directory as a plist format.
+ * Make sure to wrap this call with synchronize to support multi-threaded cases.
+ */
+- (void)saveRequests:(NSMutableArray *)requests
+{
+    if (requests.count > MAX_STORED_API_CALLS) {
+        NSRange range = NSMakeRange(requests.count - MAX_STORED_API_CALLS, MAX_STORED_API_CALLS);
+        requests = [[requests subarrayWithRange:range] mutableCopy];
+    }
+    
+    NSError *error = nil;
+    NSData *requestData = [NSPropertyListSerialization dataWithPropertyList:requests
+                                                                     format:NSPropertyListBinaryFormat_v1_0
+                                                                    options:0
+                                                                      error:&error];
+    if (requestData) {
+        [requestData writeToFile:[self documentsFilePath] atomically:YES];
+        LPLog(LPDebug, @"Saved %lu requests", (unsigned long) requests.count);
+    } else {
+        [[NSFileManager defaultManager] removeItemAtPath:[self documentsFilePath] error:&error];
+        if (error) {
+            LPLog(LPDebug, @"Error in saving requests: %@", error);
+        }
+    }
+}
 
-    return requests;
+/**
+ * Load requests from the plist in documents directory.
+ * Make sure to wrap this call with synchronize to support multi-threaded cases.
+ */
+- (NSMutableArray *)loadRequests
+{
+    @synchronized (self) {
```

Comment:
should we remove sync here too? or add it to save method again? I think it would be nice to be consistent....